doc/sphinx/user-extensions/syntax-extensions.rst

```diff
@@ -1305,6 +1305,30 @@ Displaying information about scopes
       It also displays the delimiting key if any and the class to which the
       scope is bound, if any.
 
+Impact of scopes on printing
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When several notations are available for printing the same expression,
+Coq will use the following rules for printing priorities:
+
+- If two notations are available in different scopes which are open,
+  the notation in the more recently opened scope takes precedence.
+
+- If two notations are available in the same scope, the more recently
+  defined (or imported) notation takes precedence.
+
+- Abbreviations and lonely notations, both of which have no scope,
+  take precedence over a notation in a scope if and only if the
+  abbreviation or lonely notation was defined (or imported) more
+  recently than when the corresponding scope was open, or, if the
+  scope is not open and has a delimiter, more recently than when the
+  notation in the scope was defined.
+
+- Notations which live in scopes that are closed and which have no
+  delimiters are never used for printing. A scope can be closed by
```

Comment:
@jasongross: I fixed the inaccuracy and extended the explanations. However, I realized that the new semantics adds a case when notations are chosen "arbitrarily"  by looking in a map ordered alphabetically, which basically says that the notation is chosen based on the alphabetical order of the scope in which it lives. Here is a typical example:
```
Declare Scope a.
Declare Scope c.
Declare Scope b.
Notation "#" := None : a.
Notation "###" := None : c.
Notation "##" := None : b.
Delimit Scope a with a.
Delimit Scope c with c.
Delimit Scope b with b.
Check None.
(* ### *)
```
I.e., scope `c` is chosen because last in alphabetic order, independently of the order of declarations.

This does not look satisfactory. I'll see how I can change this.

Reply:
@herbelin I am adding thus a needs progress label to this PR. Please remove it when yo consider it ready for merge consideration again.

Reply:
@jasongross: It remains to decide of a rule for ordering notations which are not in an open scope. Maybe as before, i.e. in the order of registration of notations, unless you have arguments for a different order?